stats/cloud/collector.go

```diff
@@ -196,22 +200,41 @@ func (c *Collector) Link() string {
 // at regular intervals and when the context is terminated.
 func (c *Collector) Run(ctx context.Context) {
 	wg := sync.WaitGroup{}
+	quit := make(chan struct{})
+	d := time.Duration(c.config.AggregationPeriod.Duration)
 
-	// If enabled, start periodically aggregating the collected HTTP trails
-	if c.config.AggregationPeriod.Duration > 0 {
+	switch {
+	case d == 0:
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			for {
+				select {
+				case <-c.stopSendingMetricsCh:
+					return
+				case <-ctx.Done():
+					close(quit)
+					return
+				}
+			}
+		}()
```

Comment:
I think this can basically be done with an else and `quit = ctx.Done()` ? Am I missing something ?

Reply:
You can't, because ctx.Done() return a receive only channel, thus you can't close it.

Reply:
hmm but that context is closed externally, so it should be ok?

Reply:
well he will need to have two variables .. one for the channel that is received only that is given to the for loop below to listen to and one that the aggregation goroutine will be able to close. And than set the value of the second to the first in case of aggregation ... 

I think that is better than the another goroutine approach ...

You just should be 100% certain that the first one is set to the correct value once you start waiting on it in order for there to not be some raise condition ;)

Reply:
I mean you can’t `close(ch)` where `ch` is a receive only channel.

@MStoykov I don’t get what you mean.

Reply:
The bug here due to missing ctx.Done(), because the if condition never occur if aggregation period is 0. Just we must have 2 cases to handle it, not 2 goroutine. In each case, just 1 goroutine running.

Reply:
You can have 
```golang
var quit <-chan struct{}
if aggregationOn {
   var realQuit = make(chan struct{})
   quit = realQuit
   go func() {
     // agregate here and do close(realQuit) after ctx.Done()
   })
} else {
   quit = ctx.Done()
}
```

Reply:
@MStoykov fixed.